crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+            );
+            return Ok(Some(block.block_hash()));
+        }
+
+        // if the block is stale, ignore it
+        if block.header().height < self.head_height {
+            info!(
+                block.height = %block.height(),
+                "block is stale and will not be added to the queue"
+            );
+            return Ok(None);
+        }
+
+        // if the block is at the head height OR in the future, just add it to
+        // the pending blocks
+        self.insert_to_pending_blocks(block.clone());
+
+        info!(
+            block.height = %block.height(),
+            block.hash = %block.block_hash(),
+            "block added to queue"
+        );
+        Ok(Some(block.block_hash()))
+    }
+
+    /// Removes and returns all "soft" and "Head" blocks in the queue, inorder
+    /// from oldest to newest.
+    ///
+    /// This function returns an `Option<Vec<SequencerBlockData>>`. A `Some`
+    /// value contains a vector of `SequencerBlockData` that are ready to be
+    /// passed on to execution.
+    /// A `None` value indicates that there are no blocks in the queue that are
+    /// ready to be passed on. A `None` value does not mean there are no blocks
+    /// in the queue.
+    pub(super) fn pop_blocks(&mut self) -> Option<Vec<SequencerBlockSubset>> {
+        let mut output_blocks: Vec<SequencerBlockSubset> = vec![];
+
+        let soft_blocks = self.pop_soft_blocks();
+        if let Some(mut soft_blocks) = soft_blocks {
+            output_blocks.append(soft_blocks.as_mut());
+        }
+        if let Some(mut head_blocks) = self.pop_head_blocks() {
+            output_blocks.append(head_blocks.as_mut());
+        }
+
+        if !output_blocks.is_empty() {
+            Some(output_blocks)
+        } else {
+            None
+        }
+    }
+
+    /// Removes and returns all "soft" blocks in the queue, inorder from oldest
+    /// to newest.
+    ///
+    /// This function returns an `Option<Vec<SequencerBlockData>>`. A `Some`
+    /// value contains a vector of `SequencerBlockData` that are ready to be
+    /// passed on to execution.
+    /// A `None` value indicates that there are no blocks in the queue that are
+    /// ready to be passed on. A `None` value does not mean there are no blocks
+    /// in the queue.
+    pub(super) fn pop_soft_blocks(&mut self) -> Option<Vec<SequencerBlockSubset>> {
+        let mut returned_soft_blocks: Vec<SequencerBlockSubset> =
+            self.soft_blocks.values().cloned().collect();
+        if !returned_soft_blocks.is_empty() {
+            returned_soft_blocks.sort();
+            self.soft_blocks.clear();
+            let highest_soft_block = returned_soft_blocks[returned_soft_blocks.len() - 1].clone();
+            self.head_height = highest_soft_block.height().increment();
+            self.remove_data_below_height(self.head_height);
+            Some(returned_soft_blocks)
+        } else {
+            None
+        }
+    }
```

Comment:
```suggestion
    pub(super) fn pop_soft_blocks(&mut self) -> Option<Vec<SequencerBlockSubset>> {
        if !self.soft_blocks.is_empty() {
            let returned_soft_blocks = std::mem::replace(&mut self.soft_blocks, BTreeMap::new());
            let mut returned_soft_blocks: Vec<SequencerBlockSubset> =
                returned_soft_blocks.into_values().collect();
            returned_soft_blocks.sort();
            let highest_soft_block = &returned_soft_blocks[returned_soft_blocks.len() - 1];
            self.head_height = highest_soft_block.height().increment();
            self.remove_data_below_height(self.head_height);
            Some(returned_soft_blocks)
        } else {
            None
        }
    }
```
this is probably the most efficient way to avoid allocating memory on the heap with clone. also, good to check `!self.soft_blocks.is_empty()` as early as possible to return early.